cpp/daal/src/externals/service_service_ref.h

```diff
@@ -0,0 +1,180 @@
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Declaration of service functions
+//--
+*/
+
+#ifndef __SERVICE_SERVICE_REF_H__
+#define __SERVICE_SERVICE_REF_H__
+
+#include "services/daal_defines.h"
+
+// Necessary to activate safe functions
+// But safe functions are not supported by GNU compiler
+//#define __STDC_WANT_LIB_EXT1__ 1
+
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+#include <stdio.h>
+#include <errno.h>
+
+namespace daal
+{
+namespace internal
+{
+namespace ref
+{
+struct RefService
+{
+    static void * serv_malloc(size_t size, size_t alignment) { return aligned_alloc(alignment, size); }
+
+    static void serv_free(void * ptr) { free(ptr); }
+
+    // TODO: Call of the function below in env_detect.cpp should be removed for alternative arch
+    // Remove the method after that or replace with proper call from new backend library
+    static void serv_free_buffers() {}
+
+    static int serv_memcpy_s(void * dest, size_t destSize, const void * src, size_t srcSize)
+    {
+        if (destSize < srcSize) return static_cast<int>(ENOMEM);
+        memcpy(dest, src, srcSize);
+        return 0;
+        // TODO: safe funtion
+        // return memcpy_s(dest, destSize, src, srcSize);
+    }
+
+    static int serv_memmove_s(void * dest, size_t destSize, const void * src, size_t smax)
+    {
+        if (destSize < smax) return static_cast<int>(ENOMEM);
+        memmove(dest, src, smax);
+        return 0;
+        // TODO: safe funtion
+        // return memmove_s(dest, destSize, src, smax);
+    }
+
+    static int serv_get_ht()
+    {
+        // TODO: real detection of hypertheading
+        return 0;
+    }
+
+    static int serv_get_ncpus()
+    {
+        // TODO: detection of npus
+        return 1;
+    }
+
+    static int serv_get_ncorespercpu()
+    {
+        // TODO: detection of ncores per cpu
+        return 1;
+    }
+
+    // TODO: The real call should be delegated to a backend library if the option is supported
+    static int serv_set_memory_limit(int type, size_t limit)
+    {
+        return 0;
+        // Old one - just to see what the method is for
+        // return fpk_serv_set_memory_limit(type, limit);
+    }
+
+    static size_t serv_strnlen_s(const char * src, size_t slen)
+    {
+        size_t i = 0;
+        for (; (i < slen) && (src[i] != '\0'); ++i)
+            ;
+        return i;
+    }
+
+    static int serv_strncpy_s(char * dest, size_t dmax, const char * src, size_t slen)
+    {
+        if (dmax < slen + 1) return static_cast<int>(ENOMEM);
```

Comment:
It can be root cause of some seg faults, when exception throwing
```suggestion
        if (dmax < slen) return static_cast<int>(ENOMEM);
```